lib/Transforms/IPO/StructAnalysisTransformAnalyzer.cpp

```diff
@@ -0,0 +1,865 @@
+      continue;
+    while (FieldsToTransform.size()) {
+      double MaxSum = 0;
+      FieldNumType MaxF = 0;
+      for (auto &Field : FieldsToTransform) {
+        // If the field is too large to fit in the new sub record, give up on
+        // this one
+        if (FieldSizes[Field] + NewRecordSize > Params.MaxSize)
+          continue;
+        auto Sum = 0;
+        for (auto &Fi : *NewRecord) {
+          Sum += CloseProximityRelations[Field][Fi];
+        }
+        if (Sum > MaxSum) {
+          MaxSum = Sum;
+          MaxF = Field;
+        }
+      }
+      // Take average
+      MaxSum /= NewRecord->size();
+      DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                      dbgs() << "F" << MaxF + 1
+                             << " has the best CP relations with new record: "
+                             << DEBUG_PRINT_COUNT(MaxSum) << "\n");
+      if (MaxSum < Threshold) {
+        // None of the remaining fields is suitable to put into this sub record
+        break;
+      }
+      DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT, dbgs()
+                                            << "F" << MaxF + 1
+                                            << " is added to the new record\n");
+      NewRecord->push_back(MaxF);
+      NewRecordSize += FieldSizes[MaxF];
+      FieldsToTransform.erase(MaxF);
+    }
+  }
+
+  // All the remaining fields are cold with others, create one record for all of
+  // them
+  if (FieldsToTransform.size()) {
+    auto *NewRecord = new SubRecordType;
+    SubRecords.push_back(NewRecord);
+    DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT, dbgs() << "All remaining fields are "
+                                                "cold, create one record for "
+                                                "all of them\n");
+    for (auto &F : FieldsToTransform) {
+      DEBUG_WITH_TYPE(DEBUG_TYPE_SPLIT,
+                      dbgs() << "Put F" << F + 1 << " into the cold record\n");
+      NewRecord->push_back(F);
+    }
+    FieldsToTransform.clear();
+  }
+  assert(FieldsToTransform.size() == 0);
+
+  // Print suggestions
+  outs() << "Struct definition: " << *StructureType << "\n";
+  for (unsigned i = 0; i < SubRecords.size(); i++) {
+    outs() << "Group #" << i + 1 << ":{ ";
+    for (auto &F : *SubRecords[i]) {
+      outs() << "F" << F + 1 << " ";
+    }
+    outs() << "}\n";
+  }
+}
+
+// Functions of OldFieldReorderAnalyzer. Here "old" just means the algorithm
+// used to reorder fields is derived from prior work in GCC and not necessarily
+// mean it's deprecated or worse compared to the other FieldReorderAnalyzer
+OldFieldReorderAnalyzer::OldFieldReorderAnalyzer(
+    const Module &CM, const StructType *ST, const CloseProximityBuilder *CPB,
+    const DICompositeType *DI)
+    : FieldReorderAnalyzer(CM, ST, CPB, DI, true),
+      DistanceThreshold(StructSplitDistanceThreshold) {
+  DEBUG_WITH_TYPE(DEBUG_TYPE_REORDER,
+                  dbgs() << "Initializing OldFieldReorderAnalyzer of size: "
```

Comment:
That's a good suggestion. Previously I thought the mapToDefinition() needs the CPR calculation in the constructor so I put the CPR array in it. Now I think I can safely remove them from ctor. Thanks!